sqllexer_test.go

```diff
@@ -0,0 +1,435 @@
+			input: "SELECT * FROM users where user_id = 2 and users.name = 'j'",
+			expected: []Token{
+				{IDENT, "SELECT"},
+				{WS, " "},
+				{WILDCARD, "*"},
+				{WS, " "},
+				{IDENT, "FROM"},
+				{WS, " "},
+				{IDENT, "users"},
+				{WS, " "},
+				{IDENT, "where"},
+				{WS, " "},
+				{IDENT, "user_id"},
+				{WS, " "},
+				{OPERATOR, "="},
+				{WS, " "},
+				{NUMBER, "2"},
+				{WS, " "},
+				{IDENT, "and"},
+				{WS, " "},
+				{IDENT, "users.name"},
+				{WS, " "},
+				{OPERATOR, "="},
+				{WS, " "},
+				{STRING, "'j'"},
+			},
+		},
+		{
+			name:  "select with hex and octal numbers",
+			input: "SELECT * FROM users where id = 0x123 and id = 0X123 and id = 0123",
+			expected: []Token{
+				{IDENT, "SELECT"},
+				{WS, " "},
+				{WILDCARD, "*"},
+				{WS, " "},
+				{IDENT, "FROM"},
+				{WS, " "},
+				{IDENT, "users"},
+				{WS, " "},
+				{IDENT, "where"},
+				{WS, " "},
+				{IDENT, "id"},
+				{WS, " "},
+				{OPERATOR, "="},
+				{WS, " "},
+				{NUMBER, "0x123"},
+				{WS, " "},
+				{IDENT, "and"},
+				{WS, " "},
+				{IDENT, "id"},
+				{WS, " "},
+				{OPERATOR, "="},
+				{WS, " "},
+				{NUMBER, "0X123"},
+				{WS, " "},
+				{IDENT, "and"},
+				{WS, " "},
+				{IDENT, "id"},
+				{WS, " "},
+				{OPERATOR, "="},
+				{WS, " "},
+				{NUMBER, "0123"},
+			},
+		},
+		{
+			name:  "select with float numbers and scientific notation",
+			input: "SELECT 1.2,1.2e3,1.2e-3,1.2E3,1.2E-3 FROM users",
+			expected: []Token{
+				{IDENT, "SELECT"},
+				{WS, " "},
+				{NUMBER, "1.2"},
+				{PUNCTUATION, ","},
+				{NUMBER, "1.2e3"},
+				{PUNCTUATION, ","},
+				{NUMBER, "1.2e-3"},
+				{PUNCTUATION, ","},
+				{NUMBER, "1.2E3"},
+				{PUNCTUATION, ","},
+				{NUMBER, "1.2E-3"},
+				{WS, " "},
+				{IDENT, "FROM"},
+				{WS, " "},
+				{IDENT, "users"},
+			},
+		},
+		{
+			name:  "select with double quoted identifier",
+			input: `SELECT * FROM "users table"`,
+			expected: []Token{
+				{IDENT, "SELECT"},
+				{WS, " "},
+				{WILDCARD, "*"},
+				{WS, " "},
+				{IDENT, "FROM"},
+				{WS, " "},
+				{IDENT, `"users table"`},
+			},
+		},
+		{
+			name:  "select with double quoted identifier",
+			input: `SELECT * FROM "public"."users table"`,
+			expected: []Token{
+				{IDENT, "SELECT"},
+				{WS, " "},
+				{WILDCARD, "*"},
+				{WS, " "},
+				{IDENT, "FROM"},
+				{WS, " "},
+				{IDENT, `"public"."users table"`},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			lexer := New(tt.input)
+			tokens := lexer.ScanAll()
+			if len(tokens) != len(tt.expected) {
+				t.Errorf("got %d tokens, expected %d", len(tokens), len(tt.expected))
+			}
+			for i, token := range tokens {
+				if token.Type != tt.expected[i].Type {
+					t.Errorf("got token type %d, expected %d", token.Type, tt.expected[i].Type)
+				}
+				if token.Value != tt.expected[i].Value {
+					t.Errorf("got token value %s, expected %s", token.Value, tt.expected[i].Value)
+				}
+			}
```

Comment:
I _think_ this can be done in a single assert. Either using `assert.Equal(t, tt.expected, tokens)` if we're okay with the comparison requiring the same ordering or `assert.ElementsMatch(t, tt.expected, tokens)` if the ordering isn't stable and we just want to see if the two contain the same elements regardless or order. `assert.Equal` being the nicer one because the error message is going to be so much more readable.

Reply:
Changed to `assert.Equal(t, tt.expected, tokens)` because the order of the output tokens do matter